src/traefik.py

```diff
@@ -0,0 +1,600 @@
+                "redirectScheme": {"scheme": "https", "port": 443, "permanent": True}
+            }
+
+        return {**no_prefix_middleware, **redir_scheme_middleware}
+
+    @staticmethod
+    def generate_tls_config_for_route(
+        router_name: str,
+        route_rule: str,
+        service_name: str,
+        external_host: str,
+    ) -> Dict[str, Any]:
+        """Generate a TLS configuration segment."""
+        if is_hostname(external_host):
+            tls_entry = {
+                "domains": [
+                    {
+                        "main": external_host,
+                        "sans": [f"*.{external_host}"],
+                    },
+                ],
+            }
+
+        else:
+            # When the external host is a bare IP, we do not need the 'domains' entry.
+            tls_entry = {}
+
+        return {
+            f"{router_name}-tls": {
+                "rule": route_rule,
+                "service": service_name,
+                "entryPoints": ["websecure"],
+                "tls": tls_entry,
+            }
+        }
+
+    def pull_tcp_entrypoints(self) -> Dict[str, int]:
+        """Pull the currently configured tcp entrypoints from the static config."""
+        try:
+            static_config_raw = self._container.pull(STATIC_CONFIG_PATH).read()
+        except PathError as e:
+            logger.error(f"Could not fetch static config from container; {e}")
+            return {}
+
+        static_config = yaml.safe_load(static_config_raw)
+        eps = static_config["entryPoints"]
+        return {k: v for k, v in eps.items() if k not in {"diagnostics", "web", "websecure"}}
+
+    @property
+    def is_running(self):
+        """Whether the traefik service is running."""
+        if not self._container.can_connect():
+            return False
+        return bool(self._container.get_services(self._service_name))
+
+    def restart(self):
+        """Restart the pebble service."""
+        layer = {
+            "summary": "Traefik layer",
+            "description": "Pebble config layer for Traefik",
+            "services": {
+                self._service_name: {
+                    "override": "replace",
+                    "summary": "Traefik",
+                    # trick to drop the logs to a file but also keep them available in the pod logs
+                    "command": '/bin/sh -c "{} | tee {}"'.format(BIN_PATH, LOG_PATH),
+                    "startup": "enabled",
+                },
+            },
+        }
+
+        if not self.is_running:
+            self._container.add_layer(self._layer_name, cast(LayerDict, layer), combine=True)
+            logger.debug(f"replanning {self._service_name!r} after a service update")
+            self._container.replan()
+        else:
+            logger.debug(f"restarting {self._service_name!r}")
+            self._container.restart(self._service_name)
+
+    def clear_dynamic_configs(self):
```

Comment:
We use "delete" below, so would be clearer without introducing a new term, "clear" :)
```suggestion
    def delete_all_dynamic_configs(self):
```